cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringAgentService.java

```diff
@@ -412,6 +440,62 @@ private PeerProgramUpdates getPeerProgramUpdates(String afterMessageId) {
     return new PeerProgramUpdates(peerToNotifications, lastMessageId);
   }
 
+  /**
+   * Reads from per-peer topic and returns program status updates for the given tethered peer.
+   */
+  private ImmutablePair<List<Notification>, String> getProgramStateUpdates(String peerName) {
+    String topic = programStateTopicPrefix + peerName;
+    String lastMessageId = lastMessageIdForPeer.get(peerName);
+    List<Notification> programUpdates = new ArrayList<>();
+    try (CloseableIterator<Message> iterator = messageFetcher.fetch(NamespaceId.SYSTEM.getNamespace(),
+                                                                    topic,
+                                                                    programUpdateFetchSize,
+                                                                    lastMessageIdForPeer.get(peerName))) {
+      while (iterator.hasNext()) {
+        Message message = iterator.next();
+        programUpdates.add(message.decodePayload(r -> GSON.fromJson(r, Notification.class)));
+        lastMessageId = message.getId();
+      }
+    } catch (TopicNotFoundException e) {
+      LOG.debug("Creating topic {}", topic);
```

Comment:
shall we move this log to `createTopicIfNeeded` because the topic may or may not be created at this point depending on what `createTopicIfNeeded` decides

Reply:
fixed